Automattic/jetpack
# Jetpack Search Widget: Restore support for URLs with `&post_type=post%2Cpage`

In https://github.com/Automattic/jetpack/commit/41b1919fee7b88c9718e137ca7a8b2afe361d43b#diff-5dc9f5cf9020a364a2b8c2e599d533b85062cbc85e468054bea30c823fe88956R312, `sanitize_key()` was added around `$_GET['post_type']` to improve sanitization of the user input. Unfortunately, this broke use of `&post_type=post%2Cpage` in URLs because `sanitize_key()` strips out commas.

This pull request improves sanitization of `$_GET['post_type']`, thus restoring use of commas to indicate multiple post types.

Fixes https://github.com/Automattic/wp-calypso/issues/62987

## Other information:


## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

If you don't already have one, you'll need to prepare a WordPress site with Jetpack Search active. Here's how I did so:
1. Created a fresh WordPress install in my local environment.
2. Installed Jetpack at https://github.com/Automattic/jetpack/commit/950ce392e97f88dda1871f90e5ac48a1696bd022 (via symlink to the environment).
3. Made the site publicly accessible with jurassic.tube
4. Purchased a Jetpack Search plan for the site.

Once you have a WordPress site with Jetpack Search:
1. Install and activate the Twenty Sixteen theme.
2. Disable Jetpack Instant Search:

<img width="1813" alt="image" src="https://user-images.githubusercontent.com/36432/176255978-2ee295ee-fe9c-468a-a652-300cdbc8b3dc.png">

3. Add a Legacy Jetpack Search Widget to the sidebar:

<img width="1920" alt="image" src="https://user-images.githubusercontent.com/36432/176254507-0721e013-1654-4653-a331-28ba9cdb253c.png">

4. **Important:** In the widget settings, disable searching against "Media", such that only a subset of post types are searched against:

<img width="808" alt="image" src="https://user-images.githubusercontent.com/36432/176254787-b096cb11-a4e9-4f3e-b3a5-815411853ad7.png">

5. View the Jetpack Search Widget added to the sidebar on the frontend, with `<input type="hidden" name="post_type" value="post,page">`:
 
<img width="1667" alt="image" src="https://user-images.githubusercontent.com/36432/176255035-e3f91618-cb4b-4a69-b699-867a79716a26.png">

6. Submit the search form and verify results appear when `&post_type=post%2Cpage` is present in the URL:

<img width="1323" alt="image" src="https://user-images.githubusercontent.com/36432/176255351-64968917-6345-4cc2-8b8f-405c7fced476.png">

8. Verify results also appear when the URL contains `&post_type=post` and `&post_type[]=post&post_type[]=page`:

<img width="1641" alt="image" src="https://user-images.githubusercontent.com/36432/176255665-466ced0a-cc57-419a-a0c1-271ea445ee41.png">

<img width="1300" alt="image" src="https://user-images.githubusercontent.com/36432/176255733-feb339bf-39a5-482f-a5b7-db16585d2177.png">
